### PR TITLE
Backend class, run.py refactoring

### PR DESCRIPTION
**Pull request by msehnout**
Changes (in progress)
 - [X] Abstract base class for backends
 - [X] Logger implementation for local testing
 - [x] Selinon backend stub

Other changes:
 - [X] Refactor run.py into utils
 - [X] Mock testing for utils

### CODE REVIEW

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
Since CI is dead, this is the result from `run-tests.sh`:
```
Name                           Stmts   Miss  Cover   Missing
------------------------------------------------------------
ghmonitor/__init__.py              0      0   100%
ghmonitor/backend.py              30      6    80%   15, 21, 24, 33, 36-37
ghmonitor/gopkg/__init__.py        0      0   100%
ghmonitor/gopkg/translate.py      45      2    96%   30-31
ghmonitor/models.py               68      3    96%   71-72, 74
ghmonitor/monitor.py             120     15    88%   79-86, 97-108, 128
ghmonitor/utils.py                26      1    96%   20
------------------------------------------------------------
TOTAL                            289     27    91%

Required test coverage of 40% reached. Total coverage: 90.66%
```

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
[test]

**Comment by msehnout** [comment]
@msrb I think the PR is ready for merge, I don't want it to grow even further. The selinon backend is far from ready, but there is a plenty of new tests. Pydoc and Pylint are failing because they are not set properly. See PR #7 for their fix.

File: run.py [hunk 1]
```diff
@@ -21,24 +22,18 @@
...
-                continue
-
-            if m.new_issues(new_events):
-                logger.info('There are new issues for ' + m.name)
-            if m.new_commits(new_events):
-                logger.info('There are new commits for ' + m.name)
-            if m.new_pull_requests(new_events):
-                logger.info('There are new pull requests for ' + m.name)
-            m.seen_events = new_events
+            process_new_events(monitors, backend)
```
Comment (msrb) [inline run.py]: s/monitors/m/?

File: run.py [hunk 2]
```diff
@@ -1,10 +1,11 @@
 import logging
 import os
-from time import sleep
 
+from time import sleep
 
-from ghmonitor.monitor import get_auth_header, repository_exists, get_list_of_repos
-from ghmonitor.monitor import RepositoryMonitor
+from ghmonitor.monitor import get_auth_header
+from ghmonitor.backend import LoggerBackend, create_pr_notification, create_push_notification
```
Comment (msrb) [inline run.py]: It seems like `create_pr_notification` and `create_push_notification` are not used anywhere in this module.

File: ghmonitor/monitor.py [hunk 3]
```diff
@@ -34,6 +34,16 @@ def test_get_list_of_repos():
     assert get_list_of_repos() == ['a/b', 'c/d']
 
 
+def get_list_of_packages():
```
Comment (msrb) [inline ghmonitor/monitor.py:38]: Do we need `get_list_of_repos` if we are adding `get_list_of_packages` here?

File: ghmonitor/monitor.py [hunk 4]
```diff
@@ -34,6 +34,16 @@ def test_get_list_of_repos():
...
+    repos = os.environ.get('WATCH_PACKAGES', '')
+    return repos.split(' ')
+
+
+def test_get_list_of_packages():
```
Comment (msrb) [inline ghmonitor/monitor.py]: Let's move this to ` tests/test_utils.py` so all unit tests for this module are in one place.

File: ghmonitor/utils.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+from ghmonitor.backend import create_pr_notification, create_push_notification, \
```
Comment (msrb) [inline ghmonitor/utils.py:3]: I like the following approach better, but I will leave it up to you :)
```python
from ghmonitor.backend import (
    create_pr_notification, create_push_notification, create_issue_notification
)
```

File: requirements.txt [hunk 6]
```diff
@@ -1,28 +1,79 @@
+amqp==2.1.4
```
Comment (msrb) [inline requirements.txt]: How was this file generated?

We use `pip-compile` in other repositories: https://pypi.org/project/pip-tools/

But for that we would need a `requirements.in` file here. The file would contain only direct dependencies and transitive dependencies would be injected into `requirements.txt` by `pip-compile`.

File: requirements.txt [hunk 7]
```diff
@@ -1,28 +1,79 @@
...
+pathtools==0.1.2
 pluggy==0.8.0
+psycopg2==2.7.6.1
 py==1.7.0
 pycodestyle==2.4.0
 pyee==5.0.0
+pyflakes==2.0.0
 pyppeteer==0.0.25
 pyquery==1.4.0
 pytest==4.0.1
```
Comment (msrb) [inline requirements.txt]: I think we shouldn't mix runtime and test dependencies here. I'd move test dependencies to `test-requirements.txt` under `tests/` directory.

**Comment by msrb** [comment]
@msehnout with Pavel's PR merged, please rebase on top of master and run at least following scripts:

* run-linter.sh
* check-docstyle.sh
* detect-common-errors.sh
* detect-dead-code.sh

**Comment by msrb** [comment]
Also you will need to update `cico_` scripts to build and push Docker image to the registry. See for example: https://github.com/fabric8-analytics/fabric8-analytics-github-refresh-cronjob/blob/master/cico_build_deploy.sh and https://github.com/fabric8-analytics/fabric8-analytics-github-refresh-cronjob/blob/master/cico_setup.sh.

**Comment by msrb** [comment]
Also:

```bash
$ sudo make docker-run 
docker run -e LOGLEVEL="INFO" -e GITHUB_TOKEN="" -e SLEEP_PERIOD=10 -e WATCH_REPOS="msehnout/ipc_example" github-monitor
Traceback (most recent call last):
  File "run.py", line 7, in <module>
    from ghmonitor.backend import LoggerBackend
  File "/ghmonitor/ghmonitor/backend.py", line 6, in <module>
    from f8a_worker.setup_celery import init_celery
ModuleNotFoundError: No module named 'f8a_worker'
make: *** [Makefile:11: docker-run] Error 1

```

File: run.py (hunk 1 above)
Comment (msehnout) [reply run.py]: done already, partially thanks to type hints! :-)

File: ghmonitor/monitor.py (hunk 3 above)
Comment (msehnout) [reply ghmonitor/monitor.py:38]: Not necessarily, though requirements are changing rapidly, so I decided to keep it there.

File: ghmonitor/monitor.py (hunk 4 above)
Comment (msehnout) [reply ghmonitor/monitor.py]: Well if you want to move it (I don't), then I would go with `tests/test_monitor.py`

File: requirements.txt (hunk 6 above)
Comment (msehnout) [reply requirements.txt]: ok, I created requirements.in

**Comment by msehnout** [comment]
> @msehnout with Pavel's PR merged, please rebase on top of master and run at least following scripts:
> 
>     * run-linter.sh
> 
>     * check-docstyle.sh
> 
>     * detect-common-errors.sh
> 
>     * detect-dead-code.sh

There are unused imports for typing. It is unfortunate, but Python has "comments" syntax for types, not for imports...

File: requirements.txt (hunk 7 above)
Comment (msehnout) [reply requirements.txt]: Do we really need `test-requirements.txt` file? For example `pyflakes` are installed in the `detect-common-errors.sh` script itself.

**Comment by msehnout** [comment]
> Also you will need to update `cico_` scripts to build and push Docker image to the registry. See for example: https://github.com/fabric8-analytics/fabric8-analytics-github-refresh-cronjob/blob/master/cico_build_deploy.sh and https://github.com/fabric8-analytics/fabric8-analytics-github-refresh-cronjob/blob/master/cico_setup.sh.

Not sure if I managed to do it correctly, please review again.

**Comment by msrb** [comment]
[test]

**Comment by msrb** [comment]
[test]

File: openshift/template.yaml [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+- description: Docker registry where the image is
+  displayName: Docker registry
+  required: true
+  name: DOCKER_REGISTRY
+  value: "quay.io"
+
+          - name: GITHUB_TOKEN
+          - name: SLEEP_PERIOD
+          - name: WATCH_PACKAGES
+          - name: LOGLEVEL
```
Comment (msrb) [inline openshift/template.yaml]: Looks like some leftover.

Comment (msehnout) [reply openshift/template.yaml]: Yes, it is.

**Comment by msehnout** [comment]
[test]

**Comment by msrb** [comment]
I do have few more comments, but there is already a lot of commits here. So let's merge this and continue with a new PR.

**Review by msrb — APPROVED** [review]

**Pull request merged**